art.hpp

```diff
@@ -512,122 +673,788 @@ class db final {
 
 #endif  // UNODB_DETAIL_WITH_STATS
 
-  friend auto detail::make_db_leaf_ptr<detail::node_header, db>(detail::art_key,
-                                                                value_view,
-                                                                db&);
+  friend auto detail::make_db_leaf_ptr<Key, detail::node_header, db>(
+      art_key_type, value_view, db&);
 
-  template <class, class>
+  template <class, class, template <class> class>
   friend class detail::basic_db_leaf_deleter;
 
-  template <class,                   // Db
+  template <typename,                // Key
+            template <class> class,  // Db
             template <class> class,  // CriticalSectionPolicy
             class,                   // Fake lock implementation
             class,  // Fake read_critical_section implementation
             class,  // NodePtr
-            class,  // INodeDefs
-            template <class> class,         // INodeReclamator
-            template <class, class> class>  // LeadReclamator
+            template <typename> class,         // INodeDefs
+            template <typename, class> class,  // INodeReclamator
+            template <typename, class,
+                      template <class> class>
+            class>  // LeafReclamator
   friend struct detail::basic_art_policy;
 
-  template <class, class>
+  template <class, class, template <class> class>
   friend class detail::basic_db_inode_deleter;
 
   friend struct detail::impl_helpers;
 };
 
+namespace detail {
+
+struct impl_helpers {
```

Comment:
@laurynas-biveinis I was not clear if this was a pre-existing thing which used to be elsewhere or part of the template refactor.  Either way, it's worth a comment block on what is in here and why.